src/components/TodoContext.tsx

```diff
@@ -0,0 +1,60 @@
+import React, { createContext, useCallback, useMemo, useState } from 'react';
+import { Todo } from '../types/Todo';
+import { TodosContextType } from '../types/TodosContextType';
+
+export const TodosContext = createContext<TodosContextType | undefined>(
+  undefined,
+);
+
+type Props = {
+  children: React.ReactNode;
+};
+
+export const TodosProvider: React.FC<Props> = ({ children }) => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [errorMessage, setErrorMessage] = useState('');
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [editingTodo, setEditingTodo] = useState<Todo | null>(null);
+  const [loadingTodosIds, setLoadingTodosIds] = useState<number[]>([]);
+  const [isInputFocused, setIsInputFocused] = useState(false);
+
+  const handleError = useCallback((message: string): undefined => {
+    setErrorMessage(message);
+
+    setTimeout(() => {
+      setErrorMessage('');
+    }, 3000);
+  }, []);
+
+  const todosState = useMemo(
+    () => ({
+      todos,
+      setTodos,
+      errorMessage,
+      setErrorMessage,
+      handleError,
+      tempTodo,
+      setTempTodo,
+      editingTodo,
+      setEditingTodo,
+      loadingTodosIds,
+      setLoadingTodosIds,
+      isInputFocused,
+      setIsInputFocused,
+    }),
+    [
+      todos,
+      errorMessage,
+      handleError,
+      tempTodo,
+      editingTodo,
+      loadingTodosIds,
+      isInputFocused,
+      setIsInputFocused,
+    ],
+  );
```

Comment:
The same situation with `useMemo` here. on any change of values that you included in an array react do so many useless operations.
Wrap with `useMemo/useCallback` only what you actually need. Also, then you need to wrap components where you pass cached callback with `React.memo`
https://react.dev/reference/react/useCallback